nio/client/base_client.py

```diff
@@ -129,11 +130,27 @@ class ClientCallback:
     func: Union[Callable[..., None], Callable[..., Awaitable[None]]] = field()
     filter: Union[Tuple[Type, ...], Type, None] = None
 
-    async def execute(self, event, room: Optional[MatrixRoom] = None) -> None:
+    def _execute(self, event, room: Optional[MatrixRoom] = None) -> Optional[Awaitable]:
+        # Checks the filter and executes the function once.
+        # sync_execute and async_execute will each determine
+        # how to run an awaitable if one is returned.
         if self.filter is None or isinstance(event, self.filter):
-            result = self.func(room, event) if room else self.func(event)
-            if inspect.isawaitable(result):
-                await result
+            if room:
+                return self.func(room, event)
+            else:
+                return self.func(event)
+
+    def sync_execute(self, event, room: Optional[MatrixRoom] = None) -> None:
+        """Execute callback from synchronous context."""
+        result = self._execute(event, room)
+        if inspect.isawaitable(result):
+            asyncio.run(result)
```

Comment:
I also just noticed it as I was making the doc change: `asyncio.run` has a slightly different requirement than simply taking an awaitable
```suggestion
        if inspect.iscoroutinefunction(result):
            asyncio.run(result)
```